### PR TITLE
release-23.1: cli,server: distinguish graceful vs non-graceful shutdown requests

### LINKED ISSUE #108611: cli/start: improper graceful drain upon stop trigger shutdown requests
**Describe the problem**

The shutdown logic in cli/start tries graceful drains when receiving shutdown requests. This is incorrect; currently all shutdown reasons require non-graceful shutdowns.

**Expected behavior**

For all the current reasons the code in cli/start should use a non-graceful shutdown.

Epic: CRDB-28893

Jira issue: CRDB-30548

### PR DESCRIPTION
**Pull request by knz**
Backport 1/1 commits from #108612.

/cc @cockroachdb/release

---

Fixes #108611.
Epic: CRDB-28893

---
Release justification: smoother behavior on SQL pod shutdown in Serverless

### CODE REVIEW

**Comment by yuzefovich** [comment]
I'll defer to Steven on approval for this one.

**Comment by yuzefovich** [comment]
Do we want this backported?

**Review by stevendanna — COMMENTED** [review]
Apologies for this one falling through the cracks. 

Outside of shared-process MT, do we have description of the impact of this fix for serverless?

This fix has been pretty stable on master so I'm not opposed to backporting it. Just want to make sure I correctly recall what we are fixing here.

**Comment by knz** [comment]
See the commit message for a detail of the bug;

TLDR: this patch fixes a bug whereby the top-level shutdown logic in
`cli/start.go` mistakenly attempted a graceful drain when that, in
fact, was impossible. This would show up as a storm of errors in logs
and, sometimes, block server shutdown until a non-graceful shutdown was
initiated by other means.

**Comment by ajstorm** [comment]
Does it make sense to use this backport as a test of the proposed policy changes?  Specifically, we should as the questions:

* Are all of the changes protected via a flag or option, new syntax, an environment variable or default-disabled cluster setting?
* Does this change really need to be backported? Or can it reasonably wait until the next major release to be addressed?

and, seeing as how more than 50 lines are changed:

* Is this the smallest possible change that can be made?

**Comment by knz** [comment]
@JeffSwenson has advised that CC Serverless is not currently sensitive to this so we do not need this backport.
